packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp

```diff
@@ -66,104 +66,241 @@ namespace MueLu {
+    } else {
+      Monitor m(*this, "BuildAggregatesRandom");
+      BuildAggregatesRandom(params, graph, aggregates, aggStat, numNonAggregatedNodes);
     }
 
+  } // BuildAggregates
+
+  template <class LO, class GO, class Node>
+  void AggregationPhase2bAlgorithm_kokkos<LO, GO, Node>::
+  BuildAggregatesRandom(const ParameterList& params,
+                        const LWGraph_kokkos& graph,
+                        Aggregates_kokkos& aggregates,
+                        Kokkos::View<unsigned*, typename LWGraph_kokkos::memory_space>& aggStat,
+                        LO& numNonAggregatedNodes) const {
+    using memory_space    = typename LWGraph_kokkos::memory_space;
+    using execution_space = typename LWGraph_kokkos::execution_space;
+
     const LO  numRows = graph.GetNodeNumVertices();
     const int myRank  = graph.GetComm()->getRank();
 
-    ArrayRCP<LO> vertex2AggId = aggregates.GetVertex2AggId()->getDataNonConst(0);
-    ArrayRCP<LO> procWinner   = aggregates.GetProcWinner()  ->getDataNonConst(0);
+    auto vertex2AggId           = aggregates.GetVertex2AggId()->template getLocalView<memory_space>();
+    auto procWinner             = aggregates.GetProcWinner()  ->template getLocalView<memory_space>();
+    auto colors                 = aggregates.GetGraphColors();
+    const LO numColors          = aggregates.GetGraphNumColors();
+    const LO numLocalAggregates = aggregates.GetNumAggregates();
 
-    LO numLocalAggregates = aggregates.GetNumAggregates();
+    const LO defaultConnectWeight = 100;
+    const LO penaltyConnectWeight = 10;
 
-    const int defaultConnectWeight = 100;
-    const int penaltyConnectWeight = 10;
+    Kokkos::View<LO*, memory_space> aggWeight    ("aggWeight",     numLocalAggregates);
+    Kokkos::View<LO*, memory_space> connectWeight("connectWeight", numRows);
+    Kokkos::View<LO*, memory_space> aggPenalties ("aggPenalties",  numLocalAggregates);
 
-    std::vector<int> aggWeight    (numLocalAggregates, 0);
-    std::vector<int> connectWeight(numRows, defaultConnectWeight);
-    std::vector<int> aggPenalties (numRows, 0);
+    Kokkos::deep_copy(connectWeight, defaultConnectWeight);
 
-    // We do this cycle twice.
-    // I don't know why, but ML does it too
     // taw: by running the aggregation routine more than once there is a chance that also
     // non-aggregated nodes with a node distance of two are added to existing aggregates.
     // Assuming that the aggregate size is 3 in each direction running the algorithm only twice
     // should be sufficient.
-    for (int k = 0; k < 2; k++) {
-      for (LO i = 0; i < numRows; i++) {
-        if (aggStat[i] != READY)
-          continue;
+    // lbv: If the prior phase of aggregation where run without specifying an aggregate size,
+    // the distance 2 coloring and phase 1 aggregation actually guarantee that only one iteration
+    // is needed to reach distance 2 neighbors.
+    int maxIters = 2;
```

Comment:
Would it be of interest to allow maxIters>2?  I'm not advocating adding an option now, I'm just thinking to the future.